### PR TITLE
i#1312 AVX-512 support: Fix incorrect 64-bit instr_t blackbox size.

### PR DESCRIPTION
**Pull request by hgreving2304**
Fixes an incorrect exported instr_t blackbox size.

Adds a test for above. A test for opnd_t like above should be added, however
there is currently no internal initialization method for opnd_t. opnd_t is mostly
passed around by value. We could use setter method to set bounding struct members.

Issue: #1312

### CODE REVIEW

File: suite/tests/client-interface/blackbox.c [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * DAMAGE.
+ */
+
+#include "configure.h"
+#include <stdio.h>
+
+int
+main(void)
+{
+    printf("Done\n");
```
Comment (derekbruening) [inline suite/tests/client-interface/blackbox.c]: A new file and binary are wasteful when we have a mechanism to use simple_app for client tests that don't need a custom app.  Please use that instead.

File: suite/tests/client-interface/blackbox.dll.c [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    memset(ins, 0xab, sizeof(type));       \
+    if (INSTR_LSB != 0xab) {               \
+        dr_printf("Error\n");              \
+    }                                      \
+    if (INSTR_MSB != 0xab) {               \
+        dr_printf("Error\n");              \
+    }                                      \
+    if (REDZONE != 0xcd) {                 \
+        dr_printf("Error\n");              \
+    }                                      \
```
Comment (derekbruening) [inline suite/tests/client-interface/blackbox.dll.c]: The tests up this point cannot fail: they are not testing anything AFAICT.  They will pass for *any* type.

File: suite/tests/client-interface/blackbox.dll.c [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    void *dc = dr_get_current_drcontext(); \
+    instr_init(dc, ins);                   \
+    if (INSTR_MSB != 0x0) {                \
```
Comment (derekbruening) [inline suite/tests/client-interface/blackbox.dll.c]: This assumes the final byte will be zeroed which may not be the case.  It is easy to imagine padding or a field that doesn't need up-front zeroing.

File: suite/tests/client-interface/blackbox.dll.c [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        dr_printf("Error\n");              \
+    }                                      \
+    if (REDZONE != 0xcd) {                 \
```
Comment (derekbruening) [inline suite/tests/client-interface/blackbox.dll.c]: This only tests against instr_t being larger than the black box if instr_init goes and writes to the end which may not be the case.

File: suite/tests/client-interface/blackbox.dll.c [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+#include "dr_api.h"
+#include "client_tools.h"
+#include <string.h>
+
+#define INSTR_LSB *(byte *)(&buf)
+#define INSTR_MSB *((byte *)(&buf) + sizeof(instr_t) - 1)
+#define REDZONE *((byte *)&buf + sizeof(instr_t))
+
+#define GENERATE_TEST(type)                \
+    char buf[sizeof(type) + 1];            \
```
Comment (derekbruening) [inline suite/tests/client-interface/blackbox.dll.c]: style: Use do..while for multi-statement macros to avoid the no-brace-statement problem when used.

File: suite/tests/client-interface/blackbox.dll.c [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        dr_printf("Error\n");              \
+    }                                      \
+    if (REDZONE != 0xcd) {                 \
+        dr_printf("Error\n");              \
+    }
+
+DR_EXPORT void
+dr_init(client_id_t id)
+{
+    GENERATE_TEST(instr_t)
```
Comment (derekbruening) [inline suite/tests/client-interface/blackbox.dll.c]: style: Terminating ; makes everything nicer (parser, etc.)

File: suite/tests/client-interface/blackbox.dll.c [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+}
```
Comment (derekbruening) [inline suite/tests/client-interface/blackbox.dll.c]: I was hoping for a more direct test: as mentioned offline, using the fast IR feature to get non-black-box instr_t access.  Perhaps including the headers a second time without DR_FAST_IR and `#define instr_t black_box_instr_t` or sthg, and then directly comparing the sizes.  Though still a little hacky w/ the double-include and guard define assumptions, it seems a more complete test which will detect size differences in either direction and makes zero assumptions on the field layout, padding, or on what instr_init does.

File: suite/tests/client-interface/blackbox.dll.c (hunk 3 above)
Comment (hgreving2304) [reply suite/tests/client-interface/blackbox.dll.c]: Right now, instr_init memsets the entire type. That's including padding (sizeof() includes padding).

File: suite/tests/client-interface/blackbox.dll.c (hunk 4 above)
Comment (hgreving2304) [reply suite/tests/client-interface/blackbox.dll.c]: The != 0 check above checks for instr_t being smaller than the blackbox. I am aware this test here is perfect, but it's due to the types being abstracted in the first place. By the way, did you see, there was a bug (blackbox too big).

File: suite/tests/client-interface/blackbox.dll.c (hunk 7 above)
Comment (hgreving2304) [reply suite/tests/client-interface/blackbox.dll.c]: Can look at that.

File: suite/tests/client-interface/blackbox.c (hunk 1 above)
Comment (hgreving2304) [reply suite/tests/client-interface/blackbox.c]: Ok.

File: suite/tests/client-interface/blackbox.dll.c (hunk 2 above)
Comment (hgreving2304) [reply suite/tests/client-interface/blackbox.dll.c]: That is true.

File: suite/tests/client-interface/blackbox.dll.c (hunk 3 above)
Comment (derekbruening) [reply suite/tests/client-interface/blackbox.dll.c]: Better to not make any assumptions on what instr_init does.  If we change instr_init, this test could keep passing even if we make the sizes wrong again.  My proposed method's only assumptions are on the headers, and if the assumptions fail we'll get a compile error.

File: suite/tests/client-interface/blackbox.dll.c (hunk 7 above)
Comment (derekbruening) [reply suite/tests/client-interface/blackbox.dll.c]: I guess it's unlikely to work easily: too many dup defines.  Maybe the simplest is two separate clients, one fast IR and one normal.  Could load both in the same test (or two separate runs that print, but single process is more efficient).

Comment (hgreving2304) [reply suite/tests/client-interface/blackbox.dll.c]: I had exactly the same ideas but ended up with this one here. Ack'd this test here is fragile (but better than nothing as of right now). I'll look at the double print test. Agreed that it would cover opnd_t as well.

Comment (derekbruening) [reply suite/tests/client-interface/blackbox.dll.c]: Two clients in the same process can directly compare the sizes which seems superior to printing with separate runs.

**Comment by hgreving2304** [comment]
Closing this pull request. See https://github.com/DynamoRIO/dynamorio/pull/3625.
